File: c++/Solution846.cpp
//O(nlogn) O(n), n = hand.size(), 時間複雜度為排序所需
class Solution846 {
public:
    bool isNStraightHand(vector<int>& hand, int groupSize) {
        unordered_map<int, int> cnt;
        for(int num : hand) ++cnt[num];

        sort(hand.begin(), hand.end());
        for(int num : hand){
            if(cnt[num] == 0) continue;

            for(int i=0; i<groupSize; ++i){
                if(cnt[num + i]-- == 0) return false;
            }
        }
        return true;
    }
};

File: java/Solution846.java
package leetCode.java;

import java.util.*;

//Greedy O(nlogn) O(n) 時間複雜度為排序所需
class Solution846 {
    public boolean isNStraightHand(int[] hand, int groupSize) {
        if (hand.length % groupSize != 0) {
            return false;
        }

        // 整理牌的數量
        Map<Integer, Integer> counts = new HashMap<>(); // <卡片, 數量>
        for (int num : hand) {
            counts.put(num, counts.getOrDefault(num, 0) + 1);
        }

        // 開始群組
        Arrays.sort(hand);
        for (int num : hand) {
            if (counts.get(num) == 0) {
                continue; // 卡片已經用完, 跳過
            }

            for (int i = 0; i < groupSize; ++i) {
                int target = num + i;
                if (!counts.containsKey(target) || counts.get(target) == 0) {
                    return false;
                }
                counts.put(target, counts.get(target) - 1);
            }
        }
        return true;
    }
}